### PR TITLE
fix(FEC-9523): add attach detach implemention

### PR DESCRIPTION
**Pull request by Yuvalke**
### Description of the Changes

for reloading media add attach detach implemention.
Simulate event as exist in HTML5 on destroy.

### CheckLists

- [ ] changes have been done against master branch, and PR does not conflict
- [ ] new unit / functional tests have been added (whenever applicable)
- [ ] test are passing in local environment
- [ ] Travis tests are passing (or test results are not worse than on master branch :))
- [ ] Docs have been updated

### CODE REVIEW

File: src/flash.js [hunk 1]
```diff
@@ -75,7 +75,20 @@ class Flash extends FakeEventTarget implements IEngine {
...
   _eventManager: EventManager = null;
 
-  _srcToLoad: ?string = null;
+  _source: PKMediaSourceObject;
+  /**
+   * The last time detach occurred
+   * @type {number}
+   * @private
+   */
+  _lastTimeDetach: ?number = NaN;
```
Comment (yairans) [inline src/flash.js]: `NaN` is `number`

File: src/flash.js [hunk 2]
```diff
@@ -182,12 +195,22 @@ class Flash extends FakeEventTarget implements IEngine {
...
     super();
     this._el = Utils.Dom.createElement('div');
-    this._init(source, config);
+    this.init(source, config);
   }
 
-  attachMediaSource(): void {}
+  attachMediaSource(): void {
+    this.init(this._source, this._config);
+    this._startTimeAttach = this._lastTimeDetach;
```
Comment (yairans) [inline src/flash.js]: `this. _startTimeAttach` is unnecessary. just use `this._lastTimeDetach`
see my comments https://github.com/kaltura/playkit-js/pull/425

File: src/flash.js [hunk 3]
```diff
@@ -409,8 +439,9 @@ class Flash extends FakeEventTarget implements IEngine {
       Flash._logger.warn('Missing API - Flash is not ready');
       return Promise.reject('Flash is not ready');
     }
-    this._src = this._srcToLoad;
-    this._loadPromise = this._api.load(startTime);
+    this._src = this._source ? this._source.url : null;
+    const playbackStartTime = this._startTimeAttach || startTime || 0;
```
Comment (yairans) [inline src/flash.js]: order should be `startTime || this._startTimeAttach`

File: src/flash.js [hunk 4]
```diff
@@ -197,26 +220,29 @@ class Flash extends FakeEventTarget implements IEngine {
...
+    if (this._eventManager) {
+      this._eventManager.removeAll();
+    }
+
     this._src = null;
-    this._config = null;
     this._volume = null;
     this._volumeBeforeMute = null;
-    this._srcToLoad = null;
+    this._muted = this.defaultMuted;
```
Comment (yairans) [inline src/flash.js]: related to the change? 
add explain to the description.

File: src/flashhls-adapter.js [hunk 5]
```diff
@@ -92,6 +92,13 @@ class FlashHLSAdapter extends FakeEventTarget {
     if (this._el && this._el.parentNode) {
       this._el.innerHTML = '';
     }
+    //simulate the event sequence like video tag
```
Comment (yairans) [inline src/flashhls-adapter.js]: related to the change? 
add explain to the description.

File: src/flashhls-adapter.js [hunk 6]
```diff
@@ -333,8 +352,55 @@ class FlashHLSAdapter extends FakeEventTarget {
...
+  }
+
+  /**
+   * attach media - return the media source to handle the video tag
+   * @public
+   * @returns {void}
+   */
+  attachMediaSource(): void {
+    this.attach();
+    this._startTimeAttach = this._lastTimeDetach;
```
Comment (yairans) [inline src/flashhls-adapter.js]: clean `this._startTimeAttach` on destroy (change media)

Comment (Yuvalke) [reply src/flashhls-adapter.js]: done

File: src/flashhls-adapter.js [hunk 7]
```diff
@@ -92,6 +106,17 @@ class FlashHLSAdapter extends FakeEventTarget {
     if (this._el && this._el.parentNode) {
       this._el.innerHTML = '';
     }
+    this._startTimeAttach = NaN;
+    this._lastTimeDetach = NaN;
+    delete this._api;
```
Comment (yairans) [inline src/flashhls-adapter.js]:
```suggestion
this._api = null;
```
Just change the type to `?FlashAPI`

**Pull request merged**